buffer/src/main/java/io/netty/buffer/MiMallocByteBufAllocator.java

```diff
@@ -0,0 +1,2377 @@
+        } while (use_delayed && !page.threadDelayedFreeFlag.compareAndSet(USE_DELAYED_FREE, DELAYED_FREEING));
+        // If this was the first non-local free, we need to push it on the heap delayed free list.
+        // `use_delayed` will only be true if `threadDelayedFreeFlag == USE_DELAYED_FREE`.
+        if (use_delayed) {
+            // Racy read on `heap`, but ok because `DELAYED_FREEING` is set.
+            // (see `heapCollectAbandon`)
+            LocalHeap heap = page.heapx.get();
+            if (heap != null) {
+                // Add to the delayed free list of this heap.
+                Block dfree;
+                do {
+                    dfree = heap.threadDelayedFreeList.get();
+                    block.nextBlock = dfree;
+                } while (!heap.threadDelayedFreeList.compareAndSet(dfree, block));
+            }
+            // Reset the `DELAYED_FREEING` flag.
+            boolean isReset;
+            do {
+                isReset = page.threadDelayedFreeFlag.compareAndSet(DELAYED_FREEING, NO_DELAYED_FREE);
+            } while (!isReset);
+        } else { // Common path
+            Block current;
+            do {
+                current = page.threadFreeList.get();
+                block.nextBlock = current;
+            } while (!page.threadFreeList.compareAndSet(current, block));
+        }
+    }
+
+    // Free huge block from another thread
+    private void segmentHugePageFree(Segment segment, Page page, Block block) {
+        // Huge page segments are always abandoned and can be freed immediately by any thread claim it and free.
+        LocalHeap heap = THREAD_LOCAL_HEAP.get();
+        // If this is the last reference, the CAS should always succeed.
+        if (segment.ownerThread.compareAndSet(null, Thread.currentThread())) {
+            block.nextBlock = page.freeList;
+            page.freeList = block;
+            page.usedBlocks--;
+            heap.segmentPageFree(page, true);
+        }
+    }
+
+    private void freeGenericMt(Page page, Segment segment, Block block) {
+        freeBlockMt(page, segment, block);
+    }
+
+    private void freeGenericLocal(Page page, Segment segment, Block block, LocalHeap heap) {
+        page.freeBlockLocal(block, true, heap);
+    }
+
+    /**
+     * Allocate into the given buffer. Used by {@link AdaptivePoolingAllocator.AdaptiveByteBuf#capacity(int)}.
+     */
+    void reallocate(int size, int maxCapacity, MiByteBuf into) {
+        MiByteBuf result = allocate(size, maxCapacity, into);
+        assert result == into: "Re-allocation created separate buffer instance";
+    }
+
+    ByteBuf allocate(int size, int maxCapacity) {
+        return allocate(size, maxCapacity, null);
+    }
+
+    private MiByteBuf allocate(int size, int maxCapacity, MiByteBuf byteBuf) {
+        LocalHeap localHeap = THREAD_LOCAL_HEAP.get();
+        int wSize = toWordSize(size);
+        if (size <= PAGES_FREE_DIRECT_SIZE_MAX) {
```

Comment:
The fast path here is rather different from adaptive (the size classes): it has 128 size classes distant 8 bytes each.
This enable to save the double lookup to compute the class first.

Additionally, this is always using thread local; to have a fair comparison we should enable adaptive to always use thread local as well (i made it possible at that time iirc..)

Reply:
> to have a fair comparison we should enable adaptive to always use thread local as well

The adaptive should already use thread local too in the benchmark, because the benchmark use event-loop threads, check the second param of the benchmark class constructor:
https://github.com/netty/netty/blob/e7d98900c96485026dc14feeb2a62a90695bf83c/microbench/src/main/java/io/netty/microbench/buffer/ByteBufAllocatorAllocPatternBenchmark.java#L70

Reply:
Yeah but the number of hops/indirections is still different due to the different way buffers are recycled...

Said that, my comment was more of a generic one